[#3]; feat: 게시글 작성 API 기능 추가

hubluish opened the pull request:
<!-- 이슈번호를 작성해주세요 -->
Closes #3
- Article 도메인 (Entity, Repository, Service, Controller, DTO) 구현
- 게시글 생성 시 title, content 유효성 검사 추가 (@NotBlank)
- 유효성 검사 실패 시 GlobalExceptionHandler를 통해 에러 응답 처리
- 인증된 사용자 정보(@AuthenticationPrincipal)를 활용하여 게시글 작성자 자동 연결

## 특이사항
<!-- 구현 시 고려한 사항이나 주의점이 있다면 작성해주세요 -->
- 제미나이가 짜 준 코드를 읽으면서 대충 3계층이 이런 식으로 이루어지구나 하는 감을 아주 조금이나마 잡은 것 같습니다
- 음 근데 3계층 아키텍쳐랑 MVC랑 비슷한 거라고 생각했는데 아니었군요

hubluish committed:
[#3]; feat: 게시글 도메인 초기 설정

- Article 엔티티 정의
- ArticleRepository,  ArticlerequestDto, ArticleResponseDto 정의

hubluish committed:
[#3]; feat: ArticleService 구현 및 게시글 생성 로직 추가

- ArticleService 클래스 생성
- createArticle 메서드 구현
- ArticleRequest와 사용자 이메일을 받아 게시글 생성
- 이메일을 통해 User 엔티티 조회
- Article 엔티티 생성하고 저장
- ArticleResponseDto 반환

hubluish committed:
[#3]; feat: ArticleController 구현 및 게시글 생성 API 추가

- ArticleController 클래스 생성
- POST 엔드포인트 구현
- ArticleRequest 받아 게시글 요청 처리
- @AuthenticationPrincipal 통해 로그인한 사용자 정보 획득
- ArticleService 호출해 게시글 생성 및 응답 반환

hubluish committed:
[#3]; refactor: ArticleRequestDto 유효성 검사 추가

- title과 content 필드에 @NotBlank 어노테이션 추가하여 필수 입력 값 검증

hubluish committed:
[#3]; GlobalExceptionHandler에 유효성 검사 예외 처리 추가

- MethodArgumentNotValidException 발생 시, 유효성 검사 실패 메시지를 포함한 ApiErrorResponse 반환

hubluish committed:
[#3]; test: 게시글 생성 테스트 코드 작성

- ArticleService 단위 테스트 추가 (Mockito 사용)
- ArticleController 통합 테스트 추가 (MockMvc, @WithMockUser 사용)
- 성공 및 유효성 검사 실패 케이스 검증

hubluish committed:
[#3]; fix: test 관련 오류 수정

- lombok 의존성 수정
- GlobalExceptionHandler의 APIErrorResponse 생성 방식 수정
- AuthController에 HttpStatus 추가

seulnan commented on src/main/java/com/example/backendmentoring/common/GlobalExceptionHandler.java:
```diff
@@ -1,28 +1,33 @@
-        return ApiErrorResponse.of("INTERNAL_SERVER_ERROR", "Unexpected error", req.getRequestURI());
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ApiErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex, HttpServletRequest request) {
+        String errorMessage = ex.getBindingResult().getFieldError().getDefaultMessage();
```
오 이거 아주 중요한 지점인데요?? 왜 NPE가 생기는지부터 코파일럿이 제안한 방향이 왜 해결방향인지? 공부해보면 좋을거같아요~

seulnan commented on src/main/java/com/example/backendmentoring/common/GlobalExceptionHandler.java:
```diff
@@ -1,28 +1,33 @@
-
     @ExceptionHandler(IllegalArgumentException.class)
-    @ResponseStatus(HttpStatus.BAD_REQUEST)
-    public ApiErrorResponse handleIllegalArgument(IllegalArgumentException e, HttpServletRequest req) {
-        return ApiErrorResponse.of("BAD_REQUEST", e.getMessage(), req.getRequestURI());
+    public ResponseEntity<ApiErrorResponse> handleIllegalArgumentException(IllegalArgumentException ex, HttpServletRequest request) {
+        ApiErrorResponse errorResponse = ApiErrorResponse.of(
+                HttpStatus.BAD_REQUEST.toString(),
+                ex.getMessage(),
+                request.getRequestURI()
+        );
+        return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
     }
 
-    @ExceptionHandler(Exception.class)
-    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
-    public ApiErrorResponse handleOthers(Exception e, HttpServletRequest req) {
-        return ApiErrorResponse.of("INTERNAL_SERVER_ERROR", "Unexpected error", req.getRequestURI());
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ApiErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException ex, HttpServletRequest request) {
+        String errorMessage = ex.getBindingResult().getFieldError().getDefaultMessage();
+        ApiErrorResponse errorResponse = ApiErrorResponse.of(
+                HttpStatus.BAD_REQUEST.toString(),
+                errorMessage,
+                request.getRequestURI()
+        );
+        return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
     }
```
뭐야 이것도 너무 중요한거네요 exceptionHandler가 작동하는 흐름을 이해해야합니다 자바의 예외계층도 한번 공부해보셔야하구요~

seulnan commented on src/main/java/com/example/backendmentoring/auth/AuthController.java:
```diff
@@ -3,6 +3,7 @@
 import com.example.backendmentoring.auth.dto.SignUpRequest;
 import com.example.backendmentoring.auth.dto.SignUpResponse;
 import jakarta.validation.Valid;
+import org.springframework.http.HttpStatus;
```
이건 잔소리느낌이긴한데 보통 임포트문을 보고 의존성을 파악하기때문에 안쓰는건 지우라는 의미입니다!

seulnan reviewed (approved):
고생하셨습니다. 코멘트 단건 꼭 공부해보세요.

++) 3계층과 MVC가 어떻게 다른지 느꼈다면 그 느낀점을 구체적으로 적어주시면 더 좋습니다.

hubluish committed:
refactor: article usecase 포트 도입

- CreateArticleUseCase / CreateArticleService 도입
- 컨트롤러 수정
- 4계층에 따른 아티클 도메인 파일 생성

hubluish committed:
feat: article 인프라 어댑터 추가

hubluish committed:
refactor: 추상 사용자 조회

- UserReader, UserReaderImpl 추가
- ArticleServiceTest 수정

hubluish committed:
chore: test 이름 변경

hubluish committed:
chore: 불필요한 임포트문 제거

hubluish committed:
refactor: DuplicateEmailException 추가

hubluish committed:
feat: getFieldError 추가

- null 여부 확인 가능
